taf/testlib/linux/service_lib.py

```diff
@@ -133,71 +134,133 @@ def method(service_name):
     "is_enabled",
     "is_active",
     "list",
-)
+    "daemon_reload",
+}
+
+
+def systemd_command_generator(command):
+
+    command_name = "systemctl"
+    if command in REPLACE_COMMAND_LIST:
+        command = command.replace('_', '-')
+
+    if command == "list":
+        # noinspection PyUnusedLocal
+        def list_command(_):
+            return [command_name, "list-unit-files", "--type=service"]
+        return list_command
+    elif command == "daemon-reload":
+        def daemon_reload_command(*_):
+            return [command_name, command, '']
+        return daemon_reload_command
+
+    def method(service_name):
+        return [command_name, command, "{}.service".format(service_name)]
+    return method
 
 
 class ServiceCommandGenerator(object):
 
-    def __init__(self, command_generator, command_list=COMMANDS):
+    def __getattr__(self, name):
+        if name not in self:
+            raise AttributeError(name)
+        command = self.command_generator(name)
+        setattr(self, name, command)
+        return command
+
+    def __iter__(self):
+        return iter(self.commands)
+
+    def __contains__(self, value):
+        return value in self.commands
+
+    def __init__(self, command_generator, return_codes=ReturnCodes, command_list=None):
         super(ServiceCommandGenerator, self).__init__()
+        if command_list is None:
+            command_list = COMMANDS
         self.commands = command_list
-        for command in self.commands:
-            setattr(self, command, command_generator(command))
+        self.command_generator = command_generator
+        self.return_codes = return_codes
 
 
-class SpecificServiceManager(object):
+class GenericServiceManager(object):
+    def __init__(self, run_func, command_list=None):
+        super().__init__()
+        if command_list is None:
+            command_list = COMMANDS
+        self.service_command_generator = ServiceCommandGenerator(systemd_command_generator,
+                                                                 SystemdReturnCodes,
+                                                                 command_list)
+
+        self.return_codes = SystemdReturnCodes
+        self.run_func = run_func
+
+    def __getattr__(self, name):
+        def run(service='', **kwargs):
+            return self.run_func(' '.join(command(service)), **kwargs)
```

Comment:
The run_func is given by the caller. At the point all exec_command or cli_send_command implentations in TAF expect the command to be a string. Sometimes thy even call command.split(' ') -- which is totaly wrong. (taf/testlib/clinns.py:174)
This should be fixed globaly in a separate PR.